### DISCUSSION TITLE
How to isolate data fetching inside Pinia rather than components?

### DISCUSSION THREAD

**Discussion by nathanchase** [nuxt/nuxt#16060] [Questions] [answered]
So, in **Nuxt 2**, if you had a call to dispatch an action in your Vuex store inside your `fetch()` hook...
```
async fetch() {
   await this.$store.dispatch('users/getUser', this.$route.params.id);
},
```
that you could later use `this.$fetch();` to call again... What is the **Nuxt 3** equivalent?

My workflow has been to generally have the component tell the store to do its XHR data request and then I mapped the state back to the component using `mapState`.

I have the store side of it updated and working fine in Pinia, I'm just not sure how to do the above in a component. Do I now wrap my `store.getUser()` inside of `useFetch` and then later call `refreshNuxtData` ? 

Do I need to reverse the data fetching entirely to where the component or page does the data fetch and hands it to the store, or is there a way to always have the store itself do the data fetching? I'd prefer the fetching to actually happen in the store itself, since multiple components will all call that same action to fetch the data and repopulate the store.

**Comment by nathanchase** [comment, marked as answer]
OK. I figured out how this has to work. So instead of returning the fetch in the component, you just put the action method inside a useAsyncData call. 
Example:
```js
// ./pages/user/index.vue
const { pending, data: user } = useAsyncData(username, async () => {
  await usersStore.getUser(username);  // go do your actual $fetch in the store
  return usersStore.user;  // return the value from the store's state after the fetch completes
});
```
```js
// ./stores/usersStore.ts
...
actions: {
    async getUser(username: string) {
      const response = await $fetch(`/api/users/${username}`);
      this.user = response;
    },
},
...
```
This way of having a component call a store action that does the fetch is not documented **_anywhere_** as far as I can tell, but it works. It also allows you to hook into the `pending` and `$refresh` and `refreshNuxtData` methods.

**Reply by nemtsov** [reply]
Hey @nathanchase are you still using this approach? I'm asking since [there's a decent amount of complexity with useAsyncData](https://github.com/nuxt/framework/blob/ab2456c295fc8c7609a7ef7ca1e47def5d087e87/packages/nuxt/src/app/composables/asyncData.ts#L89) to cache the value and to send it to the client for hydration, most of which pinia already does (e.g. you should be able to see the `user` (from pinia) *and* `username` (from the `useAsyncData`) in `window.__NUXT__` on the client.

I'm not certain if this is the ultimate approach, but something like this should work:

```
actions: {
    async init() {
      if (this.user) return;
      await this.getUser();
    }
...
```

And for resetting, you can either do: `userStore.$reset(); await userStore.init()` or add an `{isReset: false}` argument to init and call `userStore.init({ isReset: true })`.

**Reply by nathanchase** [reply]
Yeah, I'm still using my above approach so that I can use "pending" and "refreshNuxtData".

The only downside is that I have to use computed properties everywhere for anything that accesses a state value or getter in a store, but I think you'd have to do that with your approach as well.
